src/tla/__rewire_sequences_ext_in_apalache.tla

```diff
@@ -0,0 +1,470 @@
+
+(**
+ * TRUE iff the element e \in ToSet(s).
+ *
+ * @type: (Seq(a), a) => Bool;
+ *)
+Contains(s, e) ==
+  \E i \in DOMAIN s:
+    s[i] = e
+
+(**
+ * Reverse the given sequence s:  Let l be Len(s) (length of s).
+ * Equals a sequence s.t. << s[l], s[l - 1], ..., s[1]>>.
+ *
+ * @type: Seq(a) => Seq(a);
+ *)
+Reverse(s) ==
+  LET __s_len == Len(s) IN
+  LET __get_ith(i) == s[__s_len - i + 1] IN
+  SubSeq(__ApalacheMkSeq(__ApalacheSeqCapacity(s), __get_ith), 1, __s_len)
+
+(**
+ * The sequence s with e removed or s iff e \notin Range(s).
+ *
+ * @type: (Seq(a), a) => Seq(a);
+ *)
+Remove(seq, e) ==
+    LET __append_if_eq(__res, __t) ==
+        IF __t /= e
+        THEN Append(__res, __t)
+        ELSE __res
+    IN
+    __ApalacheFoldSeq(__append_if_eq, <<>>, seq)
+
+(**
+ * Equals the sequence s except that all occurrences of element old are
+ * replaced with the element new.
+ *
+ * @type: (Seq(a), a, a) => Seq(a);
+ *)
+ReplaceAll(seq, old, new) ==
+  LET __copy_or_set(__i) ==
+    IF seq[__i] = old THEN new ELSE seq[__i]
+  IN
+  SubSeq(__ApalacheMkSeq(__ApalacheSeqCapacity(seq), __copy_or_set),
+        1, Len(seq))
+  
+(**  
+ * Inserts element e at the position i moving the original element to i+1
+ * and so on.  In other words, a sequence t s.t.:
+ *   /\ Len(t) = Len(s) + 1
+ *   /\ t[i] = e
+ *   /\ \A j \in 1..(i - 1): t[j] = s[j]
+ *   /\ \A k \in (i + 1)..Len(s): t[k + 1] = s[k]
+ *
+ * @type: (Seq(a), Int, a) => Seq(a);
+ *)
+InsertAt(seq, k, e) ==
+  LET __copy_or_set(__i) ==
+    IF __i = k
+    THEN e
+    ELSE IF __i < k
+         THEN seq[__i]
+         ELSE seq[__i - 1]
+  IN
+  SubSeq(__ApalacheMkSeq(__ApalacheSeqCapacity(seq) + 1, __copy_or_set),
+        1, Len(seq) + 1)
+
+(**
+ * Replaces the element at position i with the element e.
+ *
+ * @type: (Seq(a), Int, a) => Seq(a);
+ *)
+ReplaceAt(s, i, e) ==
+  [s EXCEPT ![i] = e]  
+
+(**
+ * Replaces the element at position i shortening the length of s by one.
+ *
+ * @type: (Seq(a), Int) => Seq(a);
+ *)
+RemoveAt(s, i) == 
+  SubSeq(s, 1, i - 1) \o SubSeq(s, i + 1, Len(s))
+
+-----------------------------------------------------------------------------
+
+(**
+ * Cons prepends an element at the beginning of a sequence.
+ *
+ * @type: (a, Seq(a)) => Seq(a);
+ *)
+Cons(elt, seq) == 
+    <<elt>> \o seq
+
+(**
+ * The sequence formed by removing its last element.
+ *
+ * @type: Seq(a) => Seq(a);
+ *)
+Front(seq) == 
+  SubSeq(seq, 1, Len(seq) - 1)
+
+(**
+ * The last element of the sequence.
+ *
+ * @type: Seq(a) => a;
+ *)
+Last(seq) ==
+  seq[Len(seq)]
+
+-----------------------------------------------------------------------------
+
+(**
+ * TRUE iff the sequence s is a prefix of the sequence t, s.t.
+ * \E u \in Seq(Range(t)) : t = s \o u. In other words, there exists
+ * a suffix u that with s prepended equals t.
+ *
+ * @type: (Seq(a), Seq(a)) => Bool;
+ *)
+IsPrefix(s, t) ==
+  Len(s) <= Len(t) /\ SubSeq(s, 1, Len(s)) = SubSeq(t, 1, Len(s))
```

Comment:
```suggestion
  Len(s) <= Len(t) /\ s = SubSeq(t, 1, Len(s))
```

Reply:
Actually, you could probably even do 
```suggestion
  s = SubSeq(t, 1, Len(s))
```

Reply:
Good catch. We should probably propagate it to the original spec in the community modules.